File: apps/android/app/src/main/kotlin/com/epicstartup/tenant/core/localization/SiteLocale.kt
package com.epicstartup.tenant.core.localization

import com.epicstartup.tenant.core.model.PublicOrganization

/**
 * Content locales shared with the org Sites app
 * (`packages/common/src/site-locales.ts`).
 */
object SiteLocale {
	val contentLocales = listOf("en", "ar", "es", "fr", "de", "zh")

	private val labels = mapOf(
		"en" to "English",
		"ar" to "Arabic",
		"es" to "Spanish",
		"fr" to "French",
		"de" to "German",
		"zh" to "Chinese",
	)

	/** Matches `RTL_SITE_LOCALES` on the web side. */
	val rtlLocales = setOf("ar")

	fun isSupported(code: String): Boolean = contentLocales.contains(normalized(code))

	fun label(code: String): String = labels[normalized(code)] ?: code

	fun isRtl(code: String): Boolean = rtlLocales.contains(normalized(code))

	/** `"ar-SA"` → `"ar"`, `"EN"` → `"en"`. */
	fun normalized(code: String): String {
		val trimmed = code.trim().lowercase()
		if (trimmed.isEmpty()) return "en"
		return trimmed.substringBefore('-').ifEmpty { trimmed }
	}

	/**
	 * Mirrors `negotiateSiteLocale()`: exact match first, then language-only.
	 * Returns `null` when nothing matches (callers decide the fallback).
	 */
	fun match(preferred: List<String>, supported: List<String>): String? {
		val candidates = supported.map { normalized(it) }
		for (preference in preferred) {
			val exact = preference.trim().lowercase()
			if (exact.isEmpty()) continue
			val base = exact.substringBefore('-')
			val found = candidates.firstOrNull { it == exact } ?: candidates.firstOrNull { it == base }
			if (found != null) return found
		}
		return null
	}

	/**
	 * Mirrors `negotiateSiteLocale()`: exact match, then language-only, then the
	 * org default, then the first supported locale.
	 */
	fun negotiate(preferred: List<String>, supported: List<String>, defaultLocale: String): String {
		match(preferred, supported)?.let { return it }
		val candidates = supported.map(::normalized)
		val fallback = normalized(defaultLocale)
		if (candidates.contains(fallback)) return fallback
		return candidates.firstOrNull() ?: "en"
	}
}

/**
 * The language the app UI renders in.
 *
 * Resolution order:
 * 1. an explicit customer override (Profile → Language),
 * 2. the device's preferred languages, when the app has that translation,
 * 3. the locale the org negotiated for this device (the `locale` field of the
 *    `/resources/sites` payload) — so a KSA tenant still gets Arabic chrome on
 *    a phone set to a language we do not ship,
 * 4. the org's default locale, then English.
 */
data class AppLanguage(val code: String = "en") {
	val isRtl: Boolean get() = SiteLocale.isRtl(code)

	val displayName: String get() = SiteLocale.label(code)

	/** Locale tag for formatting (numbers, dates) and RTL layout direction. */
	val localeTag: String get() = code

	companion object {
		fun resolve(
			organization: PublicOrganization?,
			deviceLanguages: List<String>,
			appLocales: List<String>,
			override: String? = null,
		): AppLanguage {
			val supported = appLocales.ifEmpty { SiteLocale.contentLocales }
			val translated = supported.map { SiteLocale.normalized(it) }

			if (!override.isNullOrEmpty()) {
				return AppLanguage(
					SiteLocale.negotiate(listOf(override), supported, defaultLocale = "en"),
				)
			}

			SiteLocale.match(deviceLanguages, translated)?.let { return AppLanguage(it) }

			if (organization != null) {
				val orgMatch = SiteLocale.negotiate(
					preferred = listOf(organization.resolvedLocale, organization.resolvedDefaultLocale),
					supported = translated,
					defaultLocale = "en",
				)
				return AppLanguage(orgMatch)
			}

			return AppLanguage("en")
		}
	}
}

File: apps/ios/Sources/TenantKit/Localization/SiteLocale.swift
import Foundation

/// Content locales shared with the org Sites app (`packages/common/src/site-locales.ts`).
public enum SiteLocale {
	public static let contentLocales = ["en", "ar", "es", "fr", "de", "zh"]

	public static let labels: [String: String] = [
		"en": "English",
		"ar": "Arabic",
		"es": "Spanish",
		"fr": "French",
		"de": "German",
		"zh": "Chinese",
	]

	/// Matches `RTL_SITE_LOCALES` on the web side.
	public static let rtlLocales: Set<String> = ["ar"]

	public static func isSupported(_ code: String) -> Bool {
		contentLocales.contains(normalized(code))
	}

	public static func label(_ code: String) -> String {
		labels[normalized(code)] ?? code
	}

	public static func isRTL(_ code: String) -> Bool {
		rtlLocales.contains(normalized(code))
	}

	/// `"ar-SA"` → `"ar"`, `"EN"` → `"en"`.
	public static func normalized(_ code: String) -> String {
		let trimmed = code.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
		guard !trimmed.isEmpty else { return "en" }
		return trimmed.split(separator: "-").first.map(String.init) ?? trimmed
	}

	/// Mirrors `negotiateSiteLocale()`: exact match first, then language-only.
	/// Returns `nil` when nothing matches (callers decide the fallback).
	public static func match(preferred: [String], supported: [String]) -> String? {
		let candidates = supported.map { normalized($0) }

		for preference in preferred {
			let exact = preference.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
			guard !exact.isEmpty else { continue }
			let base = exact.split(separator: "-").first.map(String.init) ?? exact
			if let match = candidates.first(where: { $0 == exact }) ?? candidates.first(where: { $0 == base }) {
				return match
			}
		}

		return nil
	}

	/// Mirrors `negotiateSiteLocale()`: exact match, then language-only, then the
	/// org default, then the first supported locale.
	public static func negotiate(
		preferred: [String],
		supported: [String],
		defaultLocale: String
	) -> String {
		if let match = match(preferred: preferred, supported: supported) { return match }

		let candidates = supported.map { normalized($0) }
		let fallback = normalized(defaultLocale)
		if candidates.contains(fallback) { return fallback }
		return candidates.first ?? "en"
	}
}

/// The language the app UI renders in.
///
/// Resolution order:
/// 1. an explicit customer override (Profile → Language),
/// 2. the device's preferred languages, when the app has that translation,
/// 3. the locale the org negotiated for this device (the `locale` field of the
///    `/resources/sites` payload) — so a KSA tenant still gets Arabic chrome on
///    a phone set to a language we do not ship,
/// 4. the org's default locale, then English.
public struct AppLanguage: Equatable, Sendable {
	public let code: String

	public init(code: String = "en") {
		self.code = SiteLocale.normalized(code)
	}

	public var isRTL: Bool { SiteLocale.isRTL(code) }

	public var displayName: String { SiteLocale.label(code) }

	/// Locale identifier for `Foundation`/SwiftUI formatting (numbers, dates, RTL).
	public var locale: Locale { Locale(identifier: code) }

	public static func resolve(
		organization: PublicOrganization?,
		deviceLanguages: [String],
		appLocales: [String],
		override: String? = nil
	) -> AppLanguage {
		let supported = appLocales.isEmpty ? SiteLocale.contentLocales : appLocales
		let translated = supported.map { SiteLocale.normalized($0) }

		if let override, !override.isEmpty {
			return AppLanguage(
				code: SiteLocale.negotiate(preferred: [override], supported: supported, defaultLocale: "en")
			)
		}

		let deviceMatch = SiteLocale.match(preferred: deviceLanguages, supported: translated)
		if let deviceMatch {
			return AppLanguage(code: deviceMatch)
		}

		if let organization {
			let orgMatch = SiteLocale.negotiate(
				preferred: [organization.resolvedLocale, organization.resolvedDefaultLocale],
				supported: translated,
				defaultLocale: "en"
			)
			return AppLanguage(code: orgMatch)
		}

		return AppLanguage(code: "en")
	}
}
